chia/daemon/server.py

```diff
@@ -1388,30 +1359,30 @@ async def async_run_daemon(root_path: Path, wait_for_unlock: bool = False) -> in
     )
     sys.stdout.write("\n" + json_msg + "\n")
     sys.stdout.flush()
-    if lockfile is None:
+    try:
+        with Lockfile.create(daemon_launch_lock_path(root_path), timeout=1):
+            log.info(f"chia-blockchain version: {chia_full_version_str()}")
+
+            shutdown_event = asyncio.Event()
+
+            ws_server = WebSocketServer(
+                root_path,
+                ca_crt_path,
+                ca_key_path,
+                crt_path,
+                key_path,
+                shutdown_event,
+                run_check_keys_on_unlock=wait_for_unlock,
+            )
+            await ws_server.start()
+            await shutdown_event.wait()
```

Comment:
if I'm reading this correctly, the call to `wait()` here will suspend this co-routine until we're shutting down, right?
If so, this lock-file won't be released when we're done launching, but when we shut down.

it doesn't look right.

Reply:
It was the same before, wasn't it? I mean, `lockfile` still lived when the `wait()` was called or do i miss something?

Reply:
actually, it's unclear to me what `singleton` was supposed to do. It doesn't look like it's a lock file at all on win32 and cygwin. It deletes the file if it's there, and then creates a new one.

so, you might be right. But it seems to me that the *intention* was to only create a critical section for the startup, before opening the listen socket.

However, it's also not clear to me why the socket itself couldn't be used as the system-wide lock.

Reply:
> It doesn't look like it's a lock file at all on win32 and cygwin. It deletes the file if it's there, and then creates a new one.

Hm, maybe it raises the `IOError` in case the file is already locked.. should get my windows back up at some point..

> it's also not clear to me why the socket itself couldn't be used as the system-wide lock.

Hm i also wondered about that already but maybe to allow better error reporting since we know if the file is locked the daemon is running. But we don't know just from the fact that the port is bound if it's the daemon or something else unless we query something from it or so? Or is there another way to get that from the socket? Anyway, thats rather something for another PR i guess.

Reply:
I wrote this a long time ago, so I don't remember everything. But I believe there were cases where listening on an already-listened socket did *not* fail, so it was not an effective locking mechanism. It also evolved from using unix sockets where supported (ie. everything but Windows) to a tcp socket.